src/libraries/Common/src/SourceGenerators/TypeModelHelpers.cs

```diff
@@ -0,0 +1,102 @@
+
+using System;
+using System.Text;
+using Microsoft.CodeAnalysis;
+using System.Collections.Generic;
+
+namespace SourceGenerators
+{
+    internal static class TypeModelHelpers
+    {
+        private static readonly SymbolDisplayFormat s_minimalDisplayFormat = new SymbolDisplayFormat(
+            globalNamespaceStyle: SymbolDisplayGlobalNamespaceStyle.Omitted,
+            typeQualificationStyle: SymbolDisplayTypeQualificationStyle.NameAndContainingTypes,
+            genericsOptions: SymbolDisplayGenericsOptions.IncludeTypeParameters,
+            miscellaneousOptions: SymbolDisplayMiscellaneousOptions.UseSpecialTypes);
+
+        public static string ToIdentifierCompatibleSubstring(this ITypeSymbol type, bool useUniqueName = false)
+        {
+            if (type is IArrayTypeSymbol arrayType)
+            {
+                int rank = arrayType.Rank;
+                string suffix = rank == 1 ? "Array" : $"Array{rank}D"; // Array, Array2D, Array3D, ...
+                return ToIdentifierCompatibleSubstring(arrayType.ElementType) + suffix;
+            }
+
+            string name = GetName(type);
+
+            if (type is not INamedTypeSymbol namedType || !namedType.IsGenericType)
+            {
+                return name;
+            }
+
+            StringBuilder sb = new();
+
+            sb.Append(name);
+
+            foreach (ITypeSymbol genericArg in namedType.GetAllTypeArgumentsInScope())
+            {
+                sb.Append(ToIdentifierCompatibleSubstring(genericArg));
+            }
+
+            return sb.ToString();
+
+            string GetName(ITypeSymbol type) => useUniqueName
+                ? ToIdentifierCompatibleSubstring(type.ToMinimalDisplayString())
+                : type.Name;
+        }
+
+        /// <summary>
+        /// Type name, prefixed with containing type names if it is nested (e.g. ContainingType.NestedType).
+        /// </summary>
+        /// <returns></returns>
+        public static string ToMinimalDisplayString(this ITypeSymbol type) => type.ToDisplayString(s_minimalDisplayFormat);
+
+        private static ITypeSymbol[] GetAllTypeArgumentsInScope(this INamedTypeSymbol type)
+        {
+            if (!type.IsGenericType)
+            {
+                return Array.Empty<ITypeSymbol>();
+            }
+
+            var args = new List<ITypeSymbol>();
+            TraverseContainingTypes(type);
+            return args.ToArray();
+
+            void TraverseContainingTypes(INamedTypeSymbol current)
+            {
+                if (current.ContainingType is INamedTypeSymbol parent)
+                {
+                    TraverseContainingTypes(parent);
+                }
+
+                args.AddRange(current.TypeArguments);
+            }
+        }
+
+        private static string ToIdentifierCompatibleSubstring(string input)
+        {
+            StringBuilder sb = new();
+
+            foreach (char c in input)
+            {
+                if (c is '[')
```

Comment:
>if (c is '[') [](http://example.com/codeflow?start=16&length=13)

Is it possible you can get characters between `[` and `]`?

Reply:
Perhaps multi-dim arrays. I think the current code handles it okay; I'll add a test. We also have existing tests that show we skip multidim arrays as unsupported, consistent with reflection. So we shouldn't hit this codepath for multi-dim arrays.

I'll adjust the code as necessary.